domparsing/XMLSerializer-serializeToString-Element.html

```diff
@@ -0,0 +1,1139 @@
+  // -----------------------------
+
+  test(function() {
+     // ** bug (fixed) ** need to try to match the prefix (if possible, to pass and match Firefox)
+     var root = xhtml('pre:div');
+     root.setAttributeNS(XMLNSNS, "xmlns:not", HTMLNS);
+     nsPrefixDeclHTML(root).setAttributeNS(XMLNSNS, "xmlns:unused", HTMLNS);
+     eq(root, "<pre:div xmlns:not=\""+HTMLNS+"\" xmlns:pre=\""+HTMLNS+"\" xmlns:unused=\""+HTMLNS+"\"></pre:div>");
+  }, 'Duplicate namespace prefix definitions: Select matching prefix (in the middle)');
+
+  test(function() {
+     var root = xhtml('pre:div');
+     root.setAttributeNS(XMLNSNS, "xmlns:not", HTMLNS);
+     root.setAttributeNS(XMLNSNS, "xmlns:used", HTMLNS);
+     eq(root, "<used:div xmlns:not=\""+HTMLNS+"\" xmlns:used=\""+HTMLNS+"\"></used:div>");
+  }, 'Duplicate namespace prefix definitions: unmatched prefix; last prefix definition wins');
+
+  // When the element has no prefix, but is supplied with a set of available matching namespace prefix declarations, is an applicable prefix selected?
+  // -----------------------------
+
+  test(function() {
+     var root = xhtml('div');
+     root.setAttributeNS(XMLNSNS, "xmlns:not", HTMLNS);
+     root.setAttributeNS(XMLNSNS, "xmlns:used", HTMLNS);
+     eq(root, "<used:div xmlns:not=\""+HTMLNS+"\" xmlns:used=\""+HTMLNS+"\"></used:div>");
+  }, 'Duplicate namespace prefix definitions: no prefix, last prefix definition wins');
+
+  test(function() {
+  // seems like a bug to be able to do this, not legit namespace...'xml' should be reserved...
+  // CONSIDER: Dropping ns prefix decls that try to define 'xml' as a prefix, since they'll never be honored in serialization
+     var root = xml('el');
+     root.setAttributeNS(XMLNSNS, "xmlns:xml", HTMLNS);
+     root.setAttributeNS(XMLNSNS, "xmlns:used", HTMLNS);
+     eq(root, "<xml:el xmlns:xml=\""+HTMLNS+"\" xmlns:used=\""+HTMLNS+"\"/>");
+  }, 'Duplicate namespace prefix definitions: no prefix; trying to re-define xml prefix; xml prefix always wins for xml nodes');
+
+  // Muliple applicable prefixes... (ADD VARIATIONS)
+  //----------------------------
+
+  test(function() {
+     var root = nul('el');
+     root.setAttributeNS(XMLNSNS, "xmlns:a", CUSTNS);
+     root.setAttributeNS(XMLNSNS, "xmlns:b", CUSTNS);
+     root.appendChild(document.createElementNS(CUSTNS, "a:custom"));
+     root.appendChild(document.createElementNS(CUSTNS, "b:custom"));
+     eq(root, "<el xmlns:a=\""+CUSTNS+"\" xmlns:b=\""+CUSTNS+"\"><a:custom/><b:custom/></el>");
+  }, 'Multiple prefix selection');
+
+  // Multiple prefixes for same namespace (inheritance)
+  // ---------------------------
+  // html->html (delete when done with the matrix)
+
+  test(function() {
+     // ** BUG (fixed): This test shows that replacement of the prefix isn't sufficient to explain how existing implementations work. I need to keep track of all previous prefix definitions, and attempt to map them in the algorithm for optimal results.
```

Comment:
It's not clear who this comment is intended for, or what its status is at this point.